File: consumer-node/src/basePlugin.js
/**
 * Abstract base class every plugin must extend.
 *
 * Subclasses must override `get name()` and `async handle(event, settings)`.
 * The default `matches()` checks event_type against settings.event_types,
 * but plugins may override for custom eligibility logic.
 */
class BasePlugin {
  get name() {
    throw new Error("Plugin must implement the 'name' getter");
  }

  async handle(_event, _settings) {
    throw new Error("Plugin must implement handle()");
  }

  matches(event, settings) {
    const allowedTypes = (settings && settings.event_types) || [];
    if (allowedTypes.length === 0) return true;
    return allowedTypes.includes(event.event_type);
  }
}

module.exports = BasePlugin;

File: consumer-python/base_plugin.py
from abc import ABC, abstractmethod


class BasePlugin(ABC):
    """
    Abstract base class every plugin must extend.

    Subclasses provide a unique `name` and implement `handle(event)`.
    The default `matches()` checks event_type against the configured event_types list,
    but plugins can override this for custom eligibility logic.
    """

    @property
    @abstractmethod
    def name(self) -> str:
        ...

    @abstractmethod
    def handle(self, event: dict) -> None:
        """Execute the plugin action for a qualifying event."""
        ...

    def matches(self, event: dict, settings: dict) -> bool:
        """Return True if this event should be processed by this plugin."""
        allowed_types = settings.get("event_types", [])
        if not allowed_types:
            return True
        return event.get("event_type") in allowed_types
